File: engines/motion_canvas/src/utils/timeline_helper.ts
/**
 * engines/motion_canvas/src/utils/timeline_helper.ts
 * MaCode keyframe / timeline builder — TypeScript version.
 *
 * Provides interpolation and export for animation parameters.
 *
 * Usage:
 *   import {Timeline, Keyframe} from '../utils/timeline_helper';
 *
 *   const timeline = new Timeline();
 *   timeline.add(new Keyframe({t: 0.0, value: 0, ease: 'linear'}));
 *   timeline.add(new Keyframe({t: 1.0, value: 1, ease: 'ease_in_out'}));
 *   timeline.add(new Keyframe({t: 3.0, value: 0.5, ease: 'ease_out'}));
 *
 *   const v = timeline.at(1.5);   // interpolated result
 *
 *   // Export CSV (for external tools like ffmpeg)
 *   timeline.toCsv('keyframes.csv', 30);
 */

import * as fs from 'fs';

// ------------------------------------------------------------------
// 1. Easing functions
// ------------------------------------------------------------------

function easeLinear(t: number): number {
  return t;
}

function easeIn(t: number): number {
  return t * t;
}

function easeOut(t: number): number {
  return 1 - (1 - t) * (1 - t);
}

function easeInOut(t: number): number {
  if (t < 0.5) {
    return 2 * t * t;
  }
  return 1 - Math.pow(-2 * t + 2, 2) / 2;
}

function easeStep(t: number): number {
  return t < 1.0 ? 0.0 : 1.0;
}

const EASING_MAP: Record<string, (t: number) => number> = {
  linear: easeLinear,
  ease_in: easeIn,
  ease_out: easeOut,
  ease_in_out: easeInOut,
  step: easeStep,
};

function interpolate(a: number, b: number, t: number): number {
  return a + (b - a) * t;
}

// ------------------------------------------------------------------
// 2. Keyframe
// ------------------------------------------------------------------

export interface KeyframeOptions {
  t: number;
  value: number;
  ease?: string;
}

export class Keyframe {
  readonly t: number;
  readonly value: number;
  readonly ease: string;

  constructor(options: KeyframeOptions) {
    this.t = options.t;
    this.value = options.value;
    this.ease = options.ease ?? 'linear';
    if (!EASING_MAP[this.ease]) {
      throw new Error(
        `Unknown ease '${this.ease}'. Supported: ${Object.keys(EASING_MAP).join(', ')}`
      );
    }
  }
}

// ------------------------------------------------------------------
// 3. Timeline
// ------------------------------------------------------------------

export class Timeline {
  private frames: Keyframe[] = [];

  /** Add keyframe and auto-sort by time. Supports chaining. */
  add(keyframe: Keyframe): Timeline {
    this.frames.push(keyframe);
    this.frames.sort((a, b) => a.t - b.t);
    return this;
  }

  /** Clear all keyframes. */
  clear(): Timeline {
    this.frames = [];
    return this;
  }

  /**
   * Sample interpolated value at time t.
   *
   * - If t is before the first frame, returns first frame's value.
   * - If t is after the last frame, returns last frame's value.
   * - Otherwise interpolates between two frames using easing function.
   */
  at(t: number): number {
    if (this.frames.length === 0) {
      return 0.0;
    }

    if (t <= this.frames[0].t) {
      return this.frames[0].value;
    }
    if (t >= this.frames[this.frames.length - 1].t) {
      return this.frames[this.frames.length - 1].value;
    }

    for (let i = 0; i < this.frames.length - 1; i++) {
      const k0 = this.frames[i];
      const k1 = this.frames[i + 1];
      if (k0.t <= t && t <= k1.t) {
        if (k1.t === k0.t) {
          return k1.value;
        }
        const localT = (t - k0.t) / (k1.t - k0.t);
        const easedT = EASING_MAP[k0.ease](localT);
        return interpolate(k0.value, k1.value, easedT);
      }
    }

    // Fallback
    return this.frames[this.frames.length - 1].value;
  }

  /**
   * Sample timeline into CSV file.
   *
   * @param path - Output file path
   * @param fps - Sample frame rate
   * @param endT - End time; defaults to last keyframe's time
   */
  toCsv(path: string, fps = 30, endT?: number): void {
    if (this.frames.length === 0) {
      return;
    }

    const duration = endT ?? this.frames[this.frames.length - 1].t;
    const frameCount = Math.max(1, Math.ceil(duration * fps) + 1);

    const lines: string[] = ['frame,time,value'];
    for (let i = 0; i < frameCount; i++) {
      const t = i / fps;
      lines.push(`${i},${t.toFixed(6)},${this.at(t).toFixed(6)}`);
    }

    fs.writeFileSync(path, lines.join('\n'), 'utf-8');
  }

  get length(): number {
    return this.frames.length;
  }

  toString(): string {
    const frames = this.frames
      .map((k) => `(${k.t}, ${k.value}, ${k.ease})`)
      .join(', ');
    return `Timeline([${frames}])`;
  }
}

File: engines/manimgl/src/utils/timeline_helper.py
"""engines/manimgl/src/utils/timeline_helper.py
MaCode keyframe / timeline builder — provides interpolation and export for animation parameters.

Usage::

    from utils.timeline_helper import Timeline, Keyframe

    timeline = Timeline()
    timeline.add(Keyframe(t=0.0, value=0, ease="linear"))
    timeline.add(Keyframe(t=1.0, value=1, ease="ease_in_out"))
    timeline.add(Keyframe(t=3.0, value=0.5, ease="ease_out"))

    v = timeline.at(1.5)   # interpolated result

    # Export CSV (for external tools like ffmpeg)
    timeline.to_csv("keyframes.csv", fps=30)

    # Use in Manim updater
    def updater(mobject, dt):
        mobject.set_opacity(timeline.at(scene.time))
"""

import csv
import math
from collections.abc import Callable
from dataclasses import dataclass

# ------------------------------------------------------------------
# 1. Easing functions
# ------------------------------------------------------------------

def _ease_linear(t: float) -> float:
    return t


def _ease_in(t: float) -> float:
    return t * t


def _ease_out(t: float) -> float:
    return 1 - (1 - t) * (1 - t)


def _ease_in_out(t: float) -> float:
    """Smoothstep variant."""
    if t < 0.5:
        return 2 * t * t
    else:
        return 1 - math.pow(-2 * t + 2, 2) / 2


def _ease_step(t: float) -> float:
    """Step function: returns 0 if t < 1, else 1 (for discrete switching)."""
    return 0.0 if t < 1.0 else 1.0


_EASING_MAP: dict[str, Callable[[float], float]] = {
    "linear": _ease_linear,
    "ease_in": _ease_in,
    "ease_out": _ease_out,
    "ease_in_out": _ease_in_out,
    "step": _ease_step,
}


def _interpolate(a: float, b: float, t: float) -> float:
    """Linear interpolation."""
    return a + (b - a) * t


# ------------------------------------------------------------------
# 2. Keyframe
# ------------------------------------------------------------------

@dataclass
class Keyframe:
    """Single keyframe.

    Attributes:
        t: time in seconds
        value: target value (float)
        ease: easing name, see keys of ``_EASING_MAP``
    """
    t: float
    value: float
    ease: str = "linear"

    def __post_init__(self):
        if self.ease not in _EASING_MAP:
            raise ValueError(f"Unknown ease '{self.ease}'. Supported: {list(_EASING_MAP.keys())}")


# ------------------------------------------------------------------
# 3. Timeline
# ------------------------------------------------------------------

class Timeline:
    """Ordered keyframe collection, supports sampling at arbitrary times and CSV export."""

    def __init__(self):
        self._frames: list[Keyframe] = []

    def add(self, keyframe: Keyframe) -> "Timeline":
        """Add keyframe and auto-sort by time.

        Supports chaining::

            timeline.add(Keyframe(...)).add(Keyframe(...))
        """
        self._frames.append(keyframe)
        self._frames.sort(key=lambda k: k.t)
        return self

    def clear(self) -> "Timeline":
        """Clear all keyframes."""
        self._frames.clear()
        return self

    def at(self, t: float) -> float:
        """Sample interpolated value at time *t*.

        - If *t* is before the first frame, returns first frame's value.
        - If *t* is after the last frame, returns last frame's value.
        - Otherwise interpolates between two frames using easing function.

        Args:
            t: sample time in seconds

        Returns:
            float: interpolated result
        """
        if not self._frames:
            return 0.0

        if t <= self._frames[0].t:
            return self._frames[0].value
        if t >= self._frames[-1].t:
            return self._frames[-1].value

        # Find interval
        for i in range(len(self._frames) - 1):
            k0 = self._frames[i]
            k1 = self._frames[i + 1]
            if k0.t <= t <= k1.t:
                if k1.t == k0.t:
                    return k1.value
                local_t = (t - k0.t) / (k1.t - k0.t)
                eased_t = _EASING_MAP[k0.ease](local_t)
                return _interpolate(k0.value, k1.value, eased_t)

        # Fallback
        return self._frames[-1].value

    def to_csv(self, path: str, fps: int = 30, end_t: float | None = None) -> None:
        """Sample timeline into CSV file.

        Args:
            path: output file path
            fps: sample frame rate
            end_t: end time; defaults to last keyframe's time
        """
        if not self._frames:
            return

        duration = end_t if end_t is not None else self._frames[-1].t
        frame_count = max(1, int(math.ceil(duration * fps)) + 1)

        with open(path, "w", newline="", encoding="utf-8") as f:
            writer = csv.writer(f)
            writer.writerow(["frame", "time", "value"])
            for i in range(frame_count):
                t = i / fps
                writer.writerow([i, f"{t:.6f}", f"{self.at(t):.6f}"])

    def __len__(self) -> int:
        return len(self._frames)

    def __repr__(self) -> str:
        frames = ", ".join(f"({k.t}, {k.value}, {k.ease})" for k in self._frames)
        return f"Timeline([{frames}])"
